src/Processors/ResizeProcessor.cpp

```diff
@@ -430,4 +467,117 @@ IProcessor::Status StrictResizeProcessor::prepare(const PortNumbers & updated_in
     return Status::PortFull;
 }
 
+// Helper method to count how many outputs are active
+size_t MemoryDependentResizeProcessor::countActiveOutputs() const
+{
+    size_t active = 0;
+    for (size_t i = 0; i < output_ports.size(); ++i)
+    {
+        if (is_output_enabled[i] && output_ports[i].status != OutputStatus::Finished)
+            ++active;
+    }
+    return active;
+}
+
+IProcessor::Status MemoryDependentResizeProcessor::prepare(const PortNumbers & updated_inputs, const PortNumbers & updated_outputs)
+{
+    // Real definition here, or if you just want to delegate:
+    return prepareWithQueues(updated_inputs, updated_outputs);
+}
+
+Int64 MemoryDependentResizeProcessor::getFreeMemory()
```

Comment:
I do not think that `getSoftLimit` should be used here.
`getSoftLimit` is ment to be used in `GlobalOvercommitTracker`. It does not mean any useful threshold anywhere else.

I have a proposal. Lets just see on the ratios (query's memory / query hard limit, user's memory / user's hard limit, global memory / global hard limit).  
If any of them is bigger that 0.85, then we should shrink down to 1 output and do not increase the number of outputs back later.